modules/gapi/src/backends/fluid/gfluidcore_func.simd.hpp

```diff
@@ -821,6 +845,322 @@ MUL_SIMD(float, float)
+                                     v_round(vg_load_f32(&in[x + nlanes/4]) + s2)),
+                              v_pack(v_round(vg_load_f32(&in[x + nlanes/2]) + s3),
+                                     v_round(vg_load_f32(&in[x + 3*nlanes/4]) + s1))));
+
+            vx_store(&out[x + nlanes],
+                     v_pack_u(v_pack(v_round(vg_load_f32(&in[x + nlanes]) + s2),
+                                     v_round(vg_load_f32(&in[x + 5*nlanes/4]) + s3)),
+                              v_pack(v_round(vg_load_f32(&in[x + 3*nlanes/2]) + s1),
+                                     v_round(vg_load_f32(&in[x + 7*nlanes/4]) + s2))));
+
+            vx_store(&out[x + 2 * nlanes],
+                     v_pack_u(v_pack(v_round(vg_load_f32(&in[x + 2*nlanes]) + s3),
+                                     v_round(vg_load_f32(&in[x + 9*nlanes/4]) + s1)),
+                              v_pack(v_round(vg_load_f32(&in[x + 5*nlanes/2]) + s2),
+                                     v_round(vg_load_f32(&in[x + 11*nlanes/4]) + s3))));
+        }
+
+        if (x < length)
+        {
+            x = length - lanes;
+            continue;  // process unaligned tail
+        }
+        break;
+    }
+    return x;
+}
+
+//-------------------------------------------------------------------------------------------------
+
+template<typename SRC>
+CV_ALWAYS_INLINE int addc_simd_c3_impl(const SRC in[], float out[],
+                                       const v_float32& s1, const v_float32& s2,
+                                       const v_float32& s3, const int length)
+{
+    constexpr int nlanes = v_float32::nlanes;
+    constexpr int chan = 3;
+    constexpr int lanes = chan * nlanes;
+
+    int x = 0;
+    for (;;)
+    {
+        for (; x <= length - lanes; x += lanes)
+        {
+            v_float32 a1 = vg_load_f32(&in[x]);
+            v_float32 a2 = vg_load_f32(&in[x + nlanes]);
+            v_float32 a3 = vg_load_f32(&in[x + 2*nlanes]);
+
+            vx_store(&out[x], a1 + s1);
+            vx_store(&out[x + nlanes], a2 + s2);
+            vx_store(&out[x + 2*nlanes], a3 + s3);
+        }
+
+        if (x < length)
+        {
+            x = length - lanes;
+            continue;  // process unaligned tail
+        }
+        break;
+    }
+    return x;
+}
+
+//-------------------------------------------------------------------------------------------------
+
+template<typename SRC, typename DST>
+CV_ALWAYS_INLINE int addc_simd_c3(const SRC in[], const float scalar[], DST out[], const int length)
+{
+    constexpr int chan = 3;
+    constexpr int nlanes = vector_type_of_t<DST>::nlanes;
+    constexpr int lanes = chan * nlanes;
+
+    if (length < lanes)
+        return 0;
+
+    v_float32 s1 = vx_load(scalar);
+#if CV_SIMD_WIDTH == 32
+    v_float32 s2 = vx_load(&scalar[2]);
+    v_float32 s3 = vx_load(&scalar[1]);
+#else
+    v_float32 s2 = vx_load(&scalar[1]);
+    v_float32 s3 = vx_load(&scalar[2]);
```

Comment:
The difference is related to the lengths of SIMD vectors. For example let's compare SSE42 and AVX2. This function handle 3-channel case. In current snippet I load float coefficients.

SSE42 and NEON cases:
A vector has 128 bits. So I can load to it 4 32-bit float numbers. So after loading the first, the second and third vectors look like these:

First: C1 C2 C3 C1
Second: C2 C3 C1 C2
Third: C3 C1 C2 C3

As you remember I load vectors from init[] array that look like this float init[ nlanes+2 ] = {C1, C2, C3, C1, C2, C3 ...};

Therefore, in all cases except AVX2, the second vector starts loading since second init[] element (init[1] = C2) and third vector starts loading at third init[] element (init[2] = C3).

AVX2 case:
A vector has 256 bits. So I can load to it 8 32-bit floats. So 3 vectors look like these:

First: C1 C2 C3 C1 C2 C3 C1 C2
Second: C3 C1 C2 C3 C1 C2 C3 C1
Third: C2 C3 C1 C2 C3 C1 C2 C3

So, in AVX2 case the second vector starts loading since third init[] element (init[2] = C3) and third vector starts loading at second init[] element (init[1] = C2).

So, It is the reason for the difference in this snippet.

Reply:
Thank you for the explanation!